pkg/build/strategies/sti/sti.go

```diff
@@ -215,13 +215,18 @@ func (builder *STI) Build(config *api.Config) (*api.Result, error) {
 		}
 	}
 
-	if len(config.AssembleUser) > 0 {
-		glog.V(1).Infof("Running %q in %q as %q user", api.Assemble, config.Tag, config.AssembleUser)
+	assembleUser, err := builder.docker.GetAssembleUser(config.AssembleUser, config.BuilderImage)
+	if err != nil {
+		return builder.result, err
+	}
+
+	if len(assembleUser) > 0 {
+		glog.V(1).Infof("Running %q in %q as %q user", api.Assemble, config.Tag, assembleUser)
 	} else {
 		glog.V(1).Infof("Running %q in %q", api.Assemble, config.Tag)
 	}
 	startTime := time.Now()
-	if err := builder.scripts.Execute(api.Assemble, config.AssembleUser, config); err != nil {
+	if err := builder.scripts.Execute(api.Assemble, assembleUser, config); err != nil {
```

Comment:
I don't think you can change the assembleUser this late in the flow.

strategies.go:Strategy() does a "GetBuilderImage()" call and part of GetBuilderImage(via pullAndCheck() confirms that the config.AssembleUser is part of the AllowedUIDs.  So config.AssembleUser must be set to the final assemble user value before pullAndCheck is invoked, to ensure the assemble user is a legal value.